In [25]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
# %matplotlib inline

In [26]:
data=pd.read_csv("/content/StudentPerformanceFactors.csv")

In [ ]:
data.describe()

,Hours_Studied,Attendance,Sleep_Hours,Previous_Scores,Tutoring_Sessions,Physical_Activity,Exam_Score
count,6607.000000,6607.000000,6607.00000,6607.000000,6607.000000,6607.000000,6607.000000
mean,19.975329,79.977448,7.02906,75.070531,1.493719,2.967610,67.235659
std,5.990594,11.547475,1.46812,14.399784,1.230570,1.031231,3.890456
min,1.000000,60.000000,4.00000,50.000000,0.000000,0.000000,55.000000
25%,16.000000,70.000000,6.00000,63.000000,1.000000,2.000000,65.000000
50%,20.000000,80.000000,7.00000,75.000000,1.000000,3.000000,67.000000
75%,24.000000,90.000000,8.00000,88.000000,2.000000,4.000000,69.000000
max,44.000000,100.000000,10.00000,100.000000,8.000000,6.000000,101.000000


In [ ]:
data.corr(numeric_only=True)["Exam_Score"].sort_values(ascending=False)

,Exam_Score
Exam_Score,1.000000
Attendance,0.581072
Hours_Studied,0.445455
Previous_Scores,0.175079
Tutoring_Sessions,0.156525
Physical_Activity,0.027824
Sleep_Hours,-0.017022


We can dervive that as far as the Target is Exam Score, the featues affecting are Attendance,Study hours and previous score respectively.

In [ ]:
x=data[["Attendance","Hours_Studied","Previous_Scores","Tutoring_Sessions","Physical_Activity","Sleep_Hours"]] ##Features
y=data["Exam_Score"]##Target

In [ ]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
from sklearn.linear_model import LinearRegression

model=LinearRegression()
model.fit(x_train,y_train)

LinearRegression()

In [ ]:
from sklearn.metrics import r2_score

pred = model.predict(x_test)
print(r2_score(y_test, pred))

0.6416391364668392


In [ ]:
comaprison=pd.DataFrame({"Actual":y_test,"Predicted":pred})
comaprison.head()

,Actual,Predicted
743,65,66.090430
5551,65,67.255843
3442,71,68.445187
6571,64,66.765869
4204,66,65.652394


FEATURE ENGINEERING

In [ ]:
data["Study_Efficiency"]=data["Attendance"]*data["Hours_Studied"]/100

In [ ]:

data["Academic_Consistency"] = (
    data["Attendance"] + data["Previous_Scores"]
) / 2

data["Study_Sleep_Ratio"] = (
    data["Hours_Studied"] / data["Sleep_Hours"]
)

In [ ]:
data.columns

Index(['Hours_Studied', 'Attendance', 'Parental_Involvement',
       'Access_to_Resources', 'Extracurricular_Activities', 'Sleep_Hours',
       'Previous_Scores', 'Motivation_Level', 'Internet_Access',
       'Tutoring_Sessions', 'Family_Income', 'Teacher_Quality', 'School_Type',
       'Peer_Influence', 'Physical_Activity', 'Learning_Disabilities',
       'Parental_Education_Level', 'Distance_from_Home', 'Gender',
       'Exam_Score', 'Study_Efficiency', 'Academic_Consistency',
       'Study_Sleep_Ratio'],
      dtype='object')

In [ ]:
data["Motivation_Level"].unique()

array(['Low', 'Medium', 'High'], dtype=object)

In [ ]:
data["Motivation_Level"]=data["Motivation_Level"].map({"Low": 0,
                                                       "Medium":1,
                                                       "High":2})

AFTER FEATURE ENGINNERING:

In [ ]:
x=data[["Attendance","Hours_Studied","Previous_Scores","Tutoring_Sessions","Physical_Activity","Sleep_Hours","Motivation_Level","Study_Efficiency","Academic_Consistency","Study_Sleep_Ratio"]] ##Features
y=data["Exam_Score"]##Target

In [ ]:
CORR=data.corr(numeric_only=True)["Exam_Score"].sort_values(ascending=False)
CORR

,Exam_Score
Exam_Score,1.000000
Study_Efficiency,0.651652
Attendance,0.581072
Academic_Consistency,0.505110
Hours_Studied,0.445455
Study_Sleep_Ratio,0.357899
Previous_Scores,0.175079
Tutoring_Sessions,0.156525
Motivation_Level,0.086937
Physical_Activity,0.027824


In [ ]:
from sklearn.model_selection import train_test_split

# Features (Independent Variables)
X = data[
    [
        "Study_Efficiency",
        "Attendance",
        "Academic_Consistency",
        "Hours_Studied",
        "Study_Sleep_Ratio",
        "Previous_Scores",
        "Tutoring_Sessions",
        "Motivation_Level",
        "Physical_Activity",
        "Sleep_Hours"
    ]
]

# Target Variable (Dependent Variable)
y = data["Exam_Score"]

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Check shapes
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (5285, 10)
X_test shape: (1322, 10)
y_train shape: (5285,)
y_test shape: (1322,)


In [ ]:
model=LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [ ]:
from sklearn.metrics import r2_score

pred = model.predict(X_test)
print(r2_score(y_test, pred))

0.6517088923375648


Feature engineering resulted in a modest improvement in predictive performance, increasing the R² score from 0.642 to 0.652. This indicates that the engineered features captured additional patterns related to student performance.

In [29]:
from sklearn.preprocessing import MinMaxScaler
minmax = MinMaxScaler()
X_norm = minmax.fit_transform(X)

In [30]:
print("Normalized R²:", r2_score(y_test, pred))

Normalized R²: 0.6517088923375648


In [33]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [34]:
from sklearn.metrics import r2_score
print(r2_score(y_test, pred))

0.6517088923375648


**Key Insight**

Feature Engineering improved performance:

0.6517 - 0.6416 = 0.0101
Hence,accuracy improved by 0.0101 or 1.57%.


**Experiment Results:**

Built a baseline Linear Regression model.

Applied Feature Engineering by creating:

Study_Efficiency

Academic_Consistency

Study_Sleep_Ratio

Improved R² from 0.6416 → 0.6517.

Applied Min-Max Normalization and Z-Score Standardization.

Observed no significant change in model performance.